src/main/java/de/tum/in/www1/artemis/service/InternalUrlService.java

```diff
@@ -0,0 +1,108 @@
+package de.tum.in.www1.artemis.service;
+
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import de.tum.in.www1.artemis.domain.VcsRepositoryUrl;
+
+@Profile("bamboo")
+@Service
+public abstract class InternalUrlService {
+
+    protected Optional<URL> internalCiUrl;
+
+    protected Optional<URL> internalVcsUrl;
+
+    private static final Logger log = LoggerFactory.getLogger(InternalUrlService.class);
+
+    public InternalUrlService(Optional<URL> internalCiUrl, Optional<URL> internalVcsUrl) {
+        this.internalCiUrl = internalCiUrl;
+        this.internalVcsUrl = internalVcsUrl;
+    }
+
+    /**
+     * Replaces the url of the vcs repository url to the internal url if it's
+     * defined.
+     *
+     * @param vcsRepositoryUrl the vcs repository url
+     * @return the vcs repository url with the internal url
+     */
+    public VcsRepositoryUrl toInternalVcsUrl(VcsRepositoryUrl vcsRepositoryUrl) {
+        if (vcsRepositoryUrl.getURI() == null) {
+            log.warn("Cannot replace url to internal url {} because the url is null.", internalVcsUrl);
+            return vcsRepositoryUrl;
+        }
+
+        try {
+            String newInternalUrl = toInternalVcsUrl(vcsRepositoryUrl.getURI().toString());
+            return new VcsRepositoryUrl(newInternalUrl);
+        }
+        catch (URISyntaxException e) {
+            log.warn("Cannot replace url {} to {}: {}. Falling back to original url.", vcsRepositoryUrl, internalVcsUrl, e.getMessage());
+            return vcsRepositoryUrl;
+        }
+    }
+
+    /**
+     * Replaces the url of the vcs repository url to the internal url if it's
+     * defined.
+     *
+     * @param vcsRepositoryUrl the vcs repository url
+     * @return the vcs repository url with the internal url
+     */
+    public String toInternalVcsUrl(String vcsRepositoryUrl) {
+        if (internalVcsUrl.isEmpty()) {
+            return vcsRepositoryUrl;
+        }
+
+        if (vcsRepositoryUrl == null) {
+            log.warn("Cannot replace url to internal url {} because the url is null.", internalVcsUrl);
+            return vcsRepositoryUrl;
+        }
+
+        return replaceUrl(vcsRepositoryUrl, internalVcsUrl.get());
+    }
+
+    /**
+     * Replaces the url of the ci url to the internal url if it's
+     * defined.
+     *
+     * @param ciUrl the ci url
+     * @return the ci url with the internal url
+     */
+    public String toInternalCiUrl(String ciUrl) {
+        if (internalCiUrl.isPresent()) {
```

Comment:
Why is the `internalCiUrl` optional? Shouldn't it always be defined when this method is called? If yes, I think we should throw here otherwise.

Reply:
If it is not, present we don't replace the host. The internal URLs are only meant to be set in development setups where Artemis can reach Bamboo/Bitbucket via a URL but Bamboo/Bitbucket can't talk via the same one. `localhost:7990` is reachable from outside Docker from Artemis but it is not possible for Bamboo to call `localhost:7990` to check out repositories from Bitbucket.